org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionsMenu.java

```diff
@@ -84,87 +87,83 @@ public void initialize(IServiceLocator serviceLocator) {
 		CodeActionParams params = new CodeActionParams();
-		params.setTextDocument(new TextDocumentIdentifier(infos.get(0).getFileUri().toString()));
+		params.setTextDocument(new TextDocumentIdentifier(fileUri.toString()));
 		params.setRange(this.range);
 		params.setContext(context);
-		Set<CompletableFuture<?>> runningFutures = new HashSet<>();
-		for (LSPDocumentInfo info : this.infos) {
-			final CompletableFuture<List<Either<Command, CodeAction>>> codeActions = info.getInitializedLanguageClient()
-					.thenComposeAsync(languageServer -> languageServer.getTextDocumentService().codeAction(params));
-			runningFutures.add(codeActions);
-			codeActions.whenComplete((t, u) -> {
-				runningFutures.remove(codeActions);
-				UIJob job = new UIJob(menu.getDisplay(), Messages.updateCodeActions_menu) {
-					@Override
-					public IStatus runInUIThread(IProgressMonitor monitor) {
-						if (u != null) {
-							final MenuItem item = new MenuItem(menu, SWT.NONE, index);
-							item.setText(u.getMessage());
-							item.setImage(LSPImages.getSharedImage(ISharedImages.IMG_DEC_FIELD_ERROR));
-							item.setEnabled(false);
-						} else if (t != null) {
-							for (Either<Command, CodeAction> command : t) {
-								if (command != null) {
-									if (command.isLeft()) {
-										final MenuItem item = new MenuItem(menu, SWT.NONE, index);
-										item.setText(command.getLeft().getTitle());
-										item.addSelectionListener(new SelectionAdapter() {
-											@Override
-											public void widgetSelected(SelectionEvent e) {
-												executeCommand(info, command.getLeft());
-											}
-										});
-									} else if (command.isRight()) {
-										CodeAction codeAction = command.getRight();
-										final MenuItem item = new MenuItem(menu, SWT.NONE, index);
-										item.setText(codeAction.getTitle());
-										item.addSelectionListener(new SelectionAdapter() {
-											@Override
-											public void widgetSelected(SelectionEvent e) {
-												if (codeAction.getEdit() != null) {
-													LSPEclipseUtils.applyWorkspaceEdit(codeAction.getEdit(), codeAction.getTitle());
-												}
-												if (codeAction.getCommand() != null) {
-													executeCommand(info, codeAction.getCommand());
-												}
-											}
-										});
+
+		final @NonNull List<@NonNull CompletableFuture<List<Either<Command, CodeAction>>>> actions
+			= executor.computeAll((w, ls) -> ls.getTextDocumentService().codeAction(params)
+					.whenComplete((codeActions, t) -> scheduleMenuUpdate(menu, index, w, t, codeActions)));
+
+		if (actions.isEmpty()) {
+			item.setText(Messages.notImplemented);
+			return;
+		}
+		item.setText(Messages.computing);
+
+		super.fill(menu, index);
+	}
+
+	private void scheduleMenuUpdate(final Menu menu, final int index, final LanguageServerWrapper wrapper, final Throwable u, final List<Either<Command, CodeAction>> codeActions) {
```

Comment:
I like a lot that such very long lamdas as factored out as methods, but doing it as part of this PR makes hard to see the actual difference, would you mind pushing a PR with this refactor (and other similar) independently? I think we can merge those quickly and make the diff here more clear